File: packages/sdkwork-utils-php/src/DateTimeUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class DateTimeUtils
{
    public const DEFAULT_PATTERN = 'iso8601';

    public static function now(): \DateTimeImmutable
    {
        return new \DateTimeImmutable('now', new \DateTimeZone('UTC'));
    }

    public static function formatDatetime(\DateTimeImmutable $value, string $pattern = self::DEFAULT_PATTERN): string
    {
        if ($pattern !== self::DEFAULT_PATTERN) {
            throw new \InvalidArgumentException('Unsupported datetime pattern: ' . $pattern);
        }
        return $value->setTimezone(new \DateTimeZone('UTC'))->format('Y-m-d\TH:i:s.v\Z');
    }

    public static function parseDatetime(string $value, string $pattern = self::DEFAULT_PATTERN): ?\DateTimeImmutable
    {
        if ($pattern !== self::DEFAULT_PATTERN) {
            return null;
        }
        $normalized = str_replace('Z', '+00:00', trim($value));
        $parsed = \DateTimeImmutable::createFromFormat('Y-m-d\TH:i:s.vP', $normalized)
            ?: \DateTimeImmutable::createFromFormat(DATE_ATOM, $normalized);
        return $parsed?->setTimezone(new \DateTimeZone('UTC'));
    }

    public static function addDays(\DateTimeImmutable $value, int $days): \DateTimeImmutable
    {
        return $value->modify($days . ' days');
    }

    public static function addHours(\DateTimeImmutable $value, int $hours): \DateTimeImmutable
    {
        return $value->modify($hours . ' hours');
    }

    public static function addMinutes(\DateTimeImmutable $value, int $minutes): \DateTimeImmutable
    {
        return $value->modify($minutes . ' minutes');
    }

    public static function diffMillis(\DateTimeImmutable $earlier, \DateTimeImmutable $later): int
    {
        return (int) (($later->getTimestamp() - $earlier->getTimestamp()) * 1000
            + ((int) $later->format('v') - (int) $earlier->format('v')));
    }

    public static function isBefore(\DateTimeImmutable $first, \DateTimeImmutable $second): bool
    {
        return $first < $second;
    }

    public static function isAfter(\DateTimeImmutable $first, \DateTimeImmutable $second): bool
    {
        return $first > $second;
    }

    public static function startOfDayUtc(\DateTimeImmutable $value): \DateTimeImmutable
    {
        $utc = $value->setTimezone(new \DateTimeZone('UTC'));
        return $utc->setTime(0, 0, 0, 0);
    }

    public static function endOfDayUtc(\DateTimeImmutable $value): \DateTimeImmutable
    {
        return self::startOfDayUtc($value)->modify('+1 day -1 millisecond');
    }

    public static function toUnixMillis(\DateTimeImmutable $value): int
    {
        $utc = $value->setTimezone(new \DateTimeZone('UTC'));
        return $utc->getTimestamp() * 1000 + (int) $utc->format('v');
    }

    public static function fromUnixMillis(int $value): ?\DateTimeImmutable
    {
        $seconds = intdiv($value, 1000);
        $millis = abs($value % 1000);
        $parsed = \DateTimeImmutable::createFromFormat('U.v', sprintf('%d.%03d', $seconds, $millis), new \DateTimeZone('UTC'));
        return $parsed === false ? null : $parsed;
    }

    public static function isSameInstant(\DateTimeImmutable $first, \DateTimeImmutable $second): bool
    {
        return self::toUnixMillis($first) === self::toUnixMillis($second);
    }
}

File: packages/sdkwork-utils-csharp/DateTimeUtils.cs
using System.Globalization;

namespace Sdkwork.Utils;

public static class DateTimeUtils
{
    public const string DefaultPattern = "iso8601";

    public static DateTimeOffset Now() => DateTimeOffset.UtcNow;

    public static string FormatDatetime(DateTimeOffset value, string? pattern = DefaultPattern) =>
        pattern == DefaultPattern
            ? value.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'", CultureInfo.InvariantCulture)
            : throw new ArgumentException("Unsupported pattern");

    public static DateTimeOffset? ParseDatetime(string value, string? pattern = DefaultPattern) =>
        pattern == DefaultPattern
        && DateTimeOffset.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var parsed)
            ? parsed.ToUniversalTime()
            : null;

    public static DateTimeOffset AddDays(DateTimeOffset value, int days) => value.AddDays(days);
    public static DateTimeOffset AddHours(DateTimeOffset value, int hours) => value.AddHours(hours);
    public static DateTimeOffset AddMinutes(DateTimeOffset value, int minutes) => value.AddMinutes(minutes);

    public static long DiffMillis(DateTimeOffset earlier, DateTimeOffset later) =>
        (long)(later - earlier).TotalMilliseconds;

    public static bool IsBefore(DateTimeOffset first, DateTimeOffset second) => first < second;
    public static bool IsAfter(DateTimeOffset first, DateTimeOffset second) => first > second;

    public static DateTimeOffset StartOfDayUtc(DateTimeOffset value)
    {
        var utc = value.ToUniversalTime();
        return new DateTimeOffset(utc.Year, utc.Month, utc.Day, 0, 0, 0, TimeSpan.Zero);
    }

    public static DateTimeOffset EndOfDayUtc(DateTimeOffset value) =>
        StartOfDayUtc(value).AddDays(1).AddMilliseconds(-1);

    public static long ToUnixMillis(DateTimeOffset value) => value.ToUniversalTime().ToUnixTimeMilliseconds();

    public static DateTimeOffset? FromUnixMillis(long value)
    {
        try
        {
            return DateTimeOffset.FromUnixTimeMilliseconds(value).ToUniversalTime();
        }
        catch (ArgumentOutOfRangeException)
        {
            return null;
        }
    }

    public static bool IsSameInstant(DateTimeOffset first, DateTimeOffset second) =>
        first.ToUniversalTime().ToUnixTimeMilliseconds() == second.ToUniversalTime().ToUnixTimeMilliseconds();
}
